app/src/main/java/io/github/jbellis/brokk/util/BuildOutputPreprocessor.java

```diff
@@ -0,0 +1,242 @@
+
+        var result = applyCustomTimeout(
+                () -> {
+                    try {
+                        return llm.sendRequest(messages, false);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
+                    }
+                },
+                PREPROCESSING_TIMEOUT_SECONDS);
+
+        return handlePreprocessingResult(result, buildOutput, contextManager);
+    }
+
+    private static String handlePreprocessingResult(
+            Llm.StreamingResult result, String originalOutput, IContextManager contextManager) {
+        if (result.error() != null) {
+            logPreprocessingError(result.error(), contextManager);
+            return originalOutput;
+        }
+
+        String extractedErrors = result.text().trim();
+        if (extractedErrors.isBlank()) {
+            logger.warn("Build output preprocessing returned empty result. Using original output.");
+            return originalOutput;
+        }
+
+        var lines = Splitter.on('\n').splitToList(originalOutput);
+        logger.info(
+                "Successfully extracted relevant errors from build output. "
+                        + "Reduced from {} lines to {} characters.",
+                lines.size(),
+                extractedErrors.length());
+
+        return extractedErrors;
+    }
+
+    private static void logPreprocessingError(@Nullable Throwable error, IContextManager contextManager) {
+        if (error == null) {
+            logger.warn("Build output preprocessing failed with null error. Using original output.");
+            return;
+        }
+
+        boolean isCustomTimeout =
+                error.getMessage() != null && error.getMessage().contains("Build output preprocessing timed out");
+        boolean isModelTimeout =
+                error.getMessage() != null && error.getMessage().contains("timed out");
+
+        if (isCustomTimeout) {
+            logger.warn(
+                    "Build output preprocessing timed out after {} seconds (quickest model: {}). Using original output.",
+                    PREPROCESSING_TIMEOUT_SECONDS,
+                    contextManager.getService().quickestModel().getClass().getSimpleName());
+        } else if (isModelTimeout) {
+            logger.warn(
+                    "Build output preprocessing timed out at model level (quickest model: {}). Using original output.",
+                    contextManager.getService().quickestModel().getClass().getSimpleName());
+        } else {
+            logger.warn("Error during build output preprocessing: {}. Using original output.", error.getMessage());
+        }
+    }
+
+    private static SystemMessage createSystemMessage() {
+        return new SystemMessage(
+                """
+            You are familiar with common build tools (Gradle, Maven, npm, TypeScript, sbt, etc.).
```

Comment:
this needs to be generalized to apply to linting tools, i'd give one or two examples of compilers and linters at most